doc/design/scheduler.md

```diff
@@ -0,0 +1,165 @@
+# Scheduler for TrainingJob
+
+## Background
+
+We are going to define PaddlePaddle cluster job as a Kubernetes [TPR]() or
+[CRD](). Each job is described using a `yaml` representation called
+[TrainingJob](../autoscale/README.md). When a `TrainingJob` resource is
+submited to Kubernetes cluster, our customized `controller` program will
+receive an event informing the resource creation/deletion.
+
+The `controller` program should contain the following core functions:
+
+- Parser to parse `TrainingJob` resource to corresponding job components,
+  including:
+    - `ReplicaSet` of master process
+    - `ReplicaSet` or `StatefulSet` for etcd cluster
+    - `ReplicaSet` of `pserver` process
+    - `Job` of `trainer` process
+- Queue to sort `TrainingJob` resource for schedule
+- Scheduler to determine which job to run or to scale by:
+    - Job static priority
+    - Job dynamic priority
+    - Job resource request (GPU > CPU)
+    - Cluster free resource
+- Autoscaler to dynamically tunning job resources.
+
+Cases that need to be considerd during the implementaion:
+
+1. GPU is much more expensive than CPUs, jobs require GPU resource should
+have higher priority to run on GPU machines than CPU only jobs. Also a
+`TrainingJob` requires GPU must require enough CPU resource for it so that
+CPU used for laugch CUDA kernels and memory copys is not blocking the
+performance of GPU accelerated training jobs.
+1. Jobs have priorities. Some offline jobs have higher priority that can be
+able to aquire enough resource so that the job can complete at desired time,
+then job result can be updated to the production service. Other jobs like
+experiement and one-shot jobs have lower priority, they can be scaled up when
+cluster is free and can be scaled down when cluster is busy.
+1. Otherwise, jobs should share the cluster resource fairly, which means, if
```

Comment:
Topic for discussion: how "fair" do we want? If we are really fair, every jobs' trainer count will be constantly in flux, but cold starting a trainer have cost.

Maybe we can have some "freezing window", we only do the non-urgent scaling when entering the next window.

Reply:
Reasonable. The idea of "freezing window" is awesome, will add to design.